### ISSUE TITLE
Project 2 Test 4_4

### ISSUE THREAD

**Issue by kamberbatch** [nus-cs2030/2021-s1#407]
Hi, does anyone know why this test case fails and how I can resolve it? I managed to pass the public test case given on code crunch but failed this hidden test case. I'm suspecting it's because the Lazy functionality wasn't implemented properly. Pls help!
![image](https://user-images.githubusercontent.com/65482700/99406743-37b5ee80-2929-11eb-9239-fa8c39d37ab6.png)

**Comment by fans2619** [comment]
Maybe your service period is not properly generated. The nth customer arriving may not be the nth customer being served so its serving period may not be the nth number generated by the random generator.

**Comment by kamberbatch** [comment]
@fans2619 How do you implement it as such though?

**Comment by zimmmz** [comment]
I faced this issue that day too and I redesigned my whole work flow as I realized my servers are not updated properly when it goes from one event to another maybe you can check on this. And for me at first I thought I didn't do my lazy part correctly but in the end with the supplier or without the supplier the output is still the same for me so I think what matters is where you call for the random service time. And you can refer to #341 for this test case and it will return you the stats of [8.480 771 229] if your codes is right.

**Comment by qiaoningg** [comment]
Hi, it might be because you're not generating the service period at the correct place. I think it should be generated only when the customer is being served (ie in serveEvent) so only call get() on the supplier then since the nth customer arriving may not be the nth customer being served.

**Comment by kamberbatch** [comment]
@zimmmz @qiaoningg I knew I had to do that but I'm not sure how to do so. Let's say I have a Server with a MaxQueue of 2. If the server is serving a customer 1 and there is a customer 2 already in the queue. If a customer 3 comes along, wont i need to call get() on customer 2's service time in order to know when customer 3 will be served? If not, is there a better way I could implement it? Not sure if the example is clear, but I hope you get what I mean.

**Comment by qiaoningg** [comment]
What I did was that when I got an event out from my priority queue, I checked if it was a done event and if it was, I checked if the server has any waiting customer and then generate the serve event for that customer. So in this way I can know when customer 3 is served (which is when customer 2 is done being served) without calling get() on customer 2's service time.

**Comment by zimmmz** [comment]
@kamberbatch actl can I ask are you updating your servers which is oso shops at your arrive event rn?

**Comment by kamberbatch** [comment]
@qiaoningg hmm that does make sense but if thats the case how should i generate my wait event?

**Comment by kamberbatch** [comment]
@zimmmz nope. Im updating my events in EventSimulator which is a wrapper function which Main calls just to access RandomGenerator. However, I do not update the servers on ArriveEvent, only on other kinds of events.

**Comment by zimmmz** [comment]
@kamberbatch oohh cuz initially I updated my servers for the serve/wait/leave event that is return in arrive event class then after shifting it to be updated in the other respective subclasses it resolved the issue. anyways for me i only generated the random service time in the serve event sub class.

**Comment by qiaoningg** [comment]
The wait event will be generated when you execute your arrival event in Main, same as before. So in this case it's more like the serve event for a customer won't be generated by calling execute on wait event (if the customer needs to wait) but rather when your server is done serving the previous customer.

**Comment by kamberbatch** [comment]
@zimmmz hmm seems like we're doing it quite differently haha cos im not generating timings in my event subclasses

**Comment by kamberbatch** [comment]
@qiaoningg ahh sounds good let me try thanks a lot!

**Comment by kamberbatch** [comment]
@qiaoningg UPDATE: it works u saved my life thank u so much i am eternally grateful
